build_projects/dotnet-host-build/PublishTargets.cs

```diff
@@ -83,21 +85,81 @@ public static BuildTargetResult FinalizeBuild(BuildTargetContext c)
+                string flagFinalizationInProgress = $"{targetContainer}/Finalizing";
                 string semaphoreBlob = $"{Channel}/Binaries/sharedFxPublishSemaphore";
+                
                 AzurePublisherTool.CreateBlobIfNotExists(semaphoreBlob);
 
-                string leaseId = AzurePublisherTool.AcquireLeaseOnBlob(semaphoreBlob);
+                // Check if another build is executing and publishing binaries to "Latest".
+                TimeSpan maxWait = TimeSpan.FromSeconds(1800);
+                string leaseId = null;
 
-                // Prevent race conditions by dropping a version hint of what version this is. If we see this file
-                // and it is the same as our version then we know that a race happened where two+ builds finished 
-                // at the same time and someone already took care of publishing and we have no work to do.
-                if (AzurePublisherTool.IsLatestSpecifiedVersion(targetVersionFile))
+                while(true)
                 {
-                    AzurePublisherTool.ReleaseLeaseOnBlob(semaphoreBlob, leaseId);
-                    return c.Success();
+                    Stopwatch stopWatch = new Stopwatch();
+                    stopWatch.Start();
+
+                    // Wait for 30 mins to get access to finalization flag - if it does not come through, then we will
+                    // fail the build.
+                    bool fIsAnotherFinalizationInProgress = false;
+                    while (stopWatch.ElapsedMilliseconds < maxWait.TotalMilliseconds)
+                    {
+                        fIsAnotherFinalizationInProgress = AzurePublisherTool.DoesBlobReferenceExist(flagFinalizationInProgress);
+                        if (fIsAnotherFinalizationInProgress)
+                        {
+                            // Sleep for a second
+                            Thread.Sleep(TimeSpan.FromSeconds(1));
+                        }
+                        else
+                        {
+                            // No more finalization in progress
+                            break;
+                        }
+                    }
+
+                    if (fIsAnotherFinalizationInProgress)
+                    {
+                        // We timed out waiting for finalization - throw an exception
+                        throw new InvalidOperationException("Timed out while waiting for finalization flag to be cleared from another build!");
+                    }
+
+                    // Prevent race conditions by dropping a version hint of what version this is. If we see this file
+                    // and it is the same as our version then we know that a race happened where two+ builds finished 
+                    // at the same time and someone already took care of publishing and we have no work to do.
+                    if (AzurePublisherTool.DoesBlobReferenceExist(targetVersionFile))
```

Comment:
Having this outside where we have the lock makes me uncomfortable. Although in our current situation (long publishes) it wouldn't cause a race condition, if publish suddenly took less time than the `AcquireLeaseOnBlob` default timeout I think it could.